src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+
+(defvar *constant-bindings*)
+(setf (documentation '*constant-bindings* 'variable)
+      "Bound inside CONSTANT-PROPAGATION.")
+
+(defun constant-var-value (var &key no-error)
+  (the (or node null)
+       (or (cdr (assoc var *constant-bindings*))
+           (unless no-error
+             (error "Expected VAR ~S to be a constant but is not" var)))))
+
+(defun propagate-constants (node env)
+  (declare (optimize debug))
+  (labels ((constant-node-p (node)
+             ;; FIXME: Do we need more nodes classified as constants?
+             (or (node-literal-p node)
+                 (and (node-variable-p node)
+                      (tc:lookup-instance-by-codegen-sym env (node-variable-value node) :no-error t))
+                 (and (node-variable-p node)
+                      (constant-var-value (node-variable-value node) :no-error t))))
+           (constant-node-value (node &key no-error)
+             ;; FIXME: Do we need more nodes classified as constants?
+             (cond ((node-literal-p node)
+                    node)
+                   ((and (node-variable-p node)
+                         (tc:lookup-instance-by-codegen-sym env (node-variable-value node) :no-error t))
+                    node)
+                   ((node-variable-p node)
+                    (constant-var-value (node-variable-value node) :no-error no-error))
+                   (no-error
+                    nil)
+                   (t
+                    (error "Expected the node to be a constant but is not ~%~S" node))))
+
+           (propagate-constants-node-variable (node call-stack)
+             (declare (ignore call-stack)
+                      (type node-variable node))
+             (or (constant-node-value node :no-error t)
+                 node))
+
+           (propagate-constants-node-let (node call-stack)
+             (declare (type node-let node))
+             (let ((node-bindings (node-let-bindings node))
+                   (new-constant-bindings nil)
+                   (nonconstant-bindings nil))
+               (loop :for (var . value) :in node-bindings
+                     :for propagated-value-node := (funcall *traverse* value call-stack)
+                     :do (if (constant-node-p propagated-value-node)
+                             (push (cons var propagated-value-node)
+                                   new-constant-bindings)
+                             (push (cons var propagated-value-node)
+                                   nonconstant-bindings)))
+               (let ((*constant-bindings* (append new-constant-bindings *constant-bindings*)))
+                 (if nonconstant-bindings
+                     (make-node-let
+                      :type (node-type node)
+                      :bindings nonconstant-bindings
+                      :subexpr (funcall *traverse* (node-let-subexpr node) call-stack))
+                     (funcall *traverse* (node-let-subexpr node) call-stack)))))
```

Comment:
The new traversal allows arbitrary arguments to be passed through, like the `call-stack` in the current inliner and `bound-variables` in other transformations. In particular, you can use constant-bindings as an argument being passed through the traversal like this:
```suggestion
           (propagate-constants-node-let (node constant-bindings)
             (declare (type node-let node))
             (let ((node-bindings (node-let-bindings node))
                   (new-constant-bindings nil)
                   (nonconstant-bindings nil))
               (loop :for (var . value) :in node-bindings
                     :for propagated-value-node := (funcall *traverse* value constant-bindings)
                     :do (if (constant-node-p propagated-value-node)
                             (push (cons var propagated-value-node)
                                   new-constant-bindings)
                             (push (cons var propagated-value-node)
                                   nonconstant-bindings)))
               (let ((inner-constant-bindings (append new-constant-bindings constant-bindings)))
                 (if nonconstant-bindings
                     (make-node-let
                      :type (node-type node)
                      :bindings nonconstant-bindings
                      :subexpr (funcall *traverse* (node-let-subexpr node) inner-constant-bindings))
                     (funcall *traverse* (node-let-subexpr node) inner-constant-bindings)))))
```
So you don't even need the `*constant-bindings*` outside of everything unless it's better for some other reason

Reply:
Ah, I see, so the original code could have been applicable almost verbatim.

Passing along constant-bindings as a parameters does lead to slightly verbose code. But it feels more consistent with rest of the coalton codebase. 

Committed in https://github.com/coalton-lang/coalton/pull/1212/commits/ada84583b125402868f6cbe2701b7fc2def95bc8.